Add support for customising the creation of Kestrel listen sockets

deanward81 opened the pull request:
Fixes #32794. As mentioned in the issue it is currently not possible to configure the underlying listen socket used by Kestrel. In certain circumstances it is desirable to be able to configure socket options when the socket is created - a concrete case that I recently came across is setting the `SO_RECV_ANYIF` socket option on macOS so that Kestrel can listen on the `awdl0` interface (more details on that specific case [here](https://bakedbean.org.uk/posts/2021-05-airdrop-anywhere-part-3/)).

This change adds the API suggested by #32794 by adding a function that allows the caller to manage the creation of the listening socket - `Func<EndPoint, Socket> CreateListenSocket { get; set; }` allows the creation of the listen socket to be configured.

There's also some tests using IPv4, IPv6, unix domain sockets and file handles.

deanward81 committed:
Add support for configuring Kestrel listen and accept sockets

As mentioned in #32794 it is currently not possible to configure the underlying listen and accept sockets used by Kestrel. In certain circumstances it is desirable to be able to configure socket options - a concrete case that I recently came across is setting the `SO_RECV_ANYIF` socket option on macOS so that Kestrel can listen on the `awdl0` interface.

This change adds the API suggested by #32794 with some tweaks, notably splitting the configuration of the _listen_ socket from the _accept_ socket. On some platforms (*nix at least, not sure about Windows) the accept socket does not appear to inherit the socket options configured on the listen socket. So, I've added:

 - `Action<EndPoint, Socket>? ConfigureListenSocket { get; set; }` which allows the listen socket to be configured.
 - `Action<EndPoint, Socket>? ConfigureAcceptSocket { get; set; }` which allows accept sockets to be configured.

There's also some tests using IPv4, IPv6 and unix domain sockets. I have no idea how to use other kinds of `EndPoint` (e.g. `FileHandleEndPoint`) with Kestrel so have left those out of the tests. Happy to add them to get the additional coverage - just need some pointers on how to use.

Tratcher commented on src/Servers/Kestrel/Transport.Sockets/src/SocketTransportOptions.cs:
```diff
@@ -65,6 +67,16 @@ public class SocketTransportOptions
+        /// <summary>
+        /// An action used to configure an accept socket before it's passed to the underlying connection.
+        /// </summary>
+        public Action<EndPoint, Socket>? ConfigureAcceptSocket { get; set; }
```
Would `ConfigureAcceptedSocket` be clearer?

deanward81 replied:
@Tratcher no objections here, will update it tomorrow morning 👍

deanward81 committed:
`ConfigureAcceptSocket` => `ConfigureAcceptedSocket`

deanward81 committed:
Update public API bits

halter73 commented:
I wonder if it wouldn't be better to change these from `Action<EndPoint, Socket>` to `Func<EndPoint, Socket>`.  If we went this route, we'd want to expose a `Socket CreateSocket(EndPpoint)` method somewhere and document it with these callbacks.

Tratcher commented:
Create seems like a different scenario from augmentation, and requires more from the implementer.

halter73 commented on src/Servers/Kestrel/Transport.Sockets/src/SocketConnectionListener.cs:
```diff
@@ -111,14 +113,19 @@ internal void Bind()
                         listenSocket.DualMode = true;
                     }
+
+                    ConfigureSocket();
```
If we go the `Action<EndPoint, Socket>` route, wouldn't it be better to call it after binding but before listening? This would make the FileHandleEndPoint and IPEndPoint cases  more similar.

If we go the `Func<EndPoint, Socket>` route, I'd expect the returned Socket to already be bound.

halter73 commented on src/Servers/Kestrel/Transport.Sockets/src/SocketTransportOptions.cs:
```diff
@@ -65,6 +67,16 @@ public class SocketTransportOptions
+        public Action<EndPoint, Socket>? ConfigureListenSocket { get; set; }
+
+        /// <summary>
+        /// An action used to configure an accepted socket before it's passed to the underlying connection.
```
```suggestion
        /// An action used to configure an accepted socket before it's used.
```

halter73 commented:
> and requires more from the implementer.

Not really given `Socket CreateListenSocket(EndPoint)` exists.

deanward81 committed:
Update src/Servers/Kestrel/Transport.Sockets/src/SocketTransportOptions.cs

Co-authored-by: Stephen Halter <halter73@gmail.com>

JamesNK commented on src/Servers/Kestrel/Transport.Sockets/src/SocketTransportOptions.cs:
```diff
@@ -65,6 +67,16 @@ public class SocketTransportOptions
         public bool UnsafePreferInlineScheduling { get; set; }
 
+        /// <summary>
+        /// An action used to configure the listening socket before it is bound.
```
```suggestion
        /// An action used to configure the listening socket before it's bound.
```

halter73 commented:
New proposed API discussed in triage:

```diff
public class SocketTransportOptions
{
+    public Func<EndPoint, Socket> CreateListenSocket { get; set; }
+    public Socket CreateDefaultListenSocket(EndPoint);
}
```

We don't think `ConfigureAcceptedSocket` adds much value. We recommend adding [connection middleware](https://docs.microsoft.com/en-us/aspnet/core/fundamentals/servers/kestrel/endpoints?view=aspnetcore-5.0#connection-middleware) and using the new `IConnectionSocketFeature` (https://github.com/dotnet/aspnetcore/pull/31588).

davidfowl commented:
@halter73 how do you "call base" for the listen socket.

Tratcher commented:
Is CreateDefaultListenSocket supposed to be static?

davidfowl commented:
Yea it should be the stateless default.

halter73 commented:
Yeah. It should be static. I think it would basically be the current implementation [SocketConnectionListen.Bind()](https://github.com/dotnet/aspnetcore/blob/main/src/Servers/Kestrel/Transport.Sockets/src/SocketConnectionListener.cs#L87-L140), but instead of calling `Socket.Listen(int)` and setting `_listenSocket`, it would return the Socket before listening. The new implementation of `Bind()` would then call the `SocketTransportOptions.CreateListenSocket(EndPoint)` Func and then call `Socket.Listen(int)` on the returned value and set it to `_listenSocket`.

halter73 commented:
If we make it non-static, we *could* do some trickery where the SocketConnectionListener adds an internal reference to itself to SocketTransportOptions before calling the CreateDefaultListenSocket callback allowing CreateDefaultListenSocket to set `_socketHandle` if need be. Not sure if this would be worthwhile, but I do like that it makes it work in all cases.

deanward81 commented:
> We might have to bypass the CreateListenSocket for FileHandleEndPoints since I can't think of a great way to manage the `_socketHandle` in that case. I think this is fine because if you're using a FileHandleEndPoint, you can configure the socket options before passing it to `KestrelServerOptions.ListenHandle(long)`.

Does the `_socketHandle` field need to be there? It's only used during disposal, but that handle is created with `ownsHandle: true` which, when passed to `Socket`, means the handle is cleaned up in the finaliser of the socket (which arguably is more correct because it's an unmanaged resource). https://github.com/dotnet/runtime/blob/main/src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.cs#L3262

That said, is the handle _really_ owned in this case? The user is passing an opaque handle created elsewhere - is it intentional that it is marked as "owned" by the socket transport? Shouldn't they be responsible for cleanup?

Reason I ask - not having that field allows `CreateDefaultListenSocket` to work for all cases and keeps things decoupled (no need to do mess with references to the `SocketConnectionListener` in the instance case).

deanward81 committed:
Move to using API from triage

This removes `ConfigureAcceptedSocket` and changes `ConfigureListenSocket` to be a factory for the socket. A static method that creates the default socket is defined in `SocketTransportOptions.CreateDefaultListenSocket` - this effectively lifts the code that created a socket for an `EndPoint` in `SocketConnectionListener.Bind` to `SocketTransportOptions`.

If `SocketTransportOptions.CreateListenSocket` is set then it is used in preference of `SocketTransportOptions.CreateDefaultListenSocket` and it is expected that the function creates the right type of socket for the passed `EndPoint`. Implementors can call `SocketTransportOptions.CreateDefaultListenSocket` themselves and manipulate the returned socket instance as they see fit.

Note that during implementation I removed the `_socketHandle` field from `SocketConnectionListener` - this was only set so that `Dispose` could be called when the listener is disposed. Under the hood `Socket` already disposes a handle passed to it during finalization, but only if the `ownsHandle` parameter is `true` . In this case the `SafeSocketHandle` _is_ instantiated with  this parameter so the the underlying handle will be closed when the `_listenSocket` field is disposed - that is currently the case when the listener is disposed.

deanward81 commented:
I've updated to use the triaged API (using a static `CreateDefaultListenSocket`). I removed the `_listenHandle` in the last commit and, so far as I can tell, there are no negative consequences (tests pass, at least). Details (from the commit):

> Note that during implementation I removed the `_socketHandle` field from `SocketConnectionListener` - this was only set so that `Dispose` could be called when the listener is disposed. Under the hood `Socket` already disposes a handle passed to it during finalization, but only if the `ownsHandle` parameter is `true` . In this case the `SafeSocketHandle` _is_ instantiated with  this parameter so the the underlying handle will be closed when the `_listenSocket` field is disposed - that is currently the case when the listener is disposed.

halter73 commented on src/Servers/Kestrel/Transport.Sockets/src/SocketConnectionListener.cs:
```diff
@@ -23,7 +23,6 @@ internal sealed class SocketConnectionListener : IConnectionListener
         private Socket? _listenSocket;
         private int _settingsIndex;
         private readonly SocketTransportOptions _options;
-        private SafeSocketHandle? _socketHandle;
```
I think it's fine to remove this since we probably shouldn't have taken ownership of the handle in the first place, but I want to point out to other reviewers this is a behavioral change.

halter73 commented on src/Servers/Kestrel/Transport.Sockets/src/SocketTransportOptions.cs:
```diff
@@ -65,6 +68,47 @@ public class SocketTransportOptions
+        /// A function used to create a new <see cref="Socket"/> to listen with. If
+        /// not set, <see cref="CreateDefaultListenSocket" /> is called instead.
+        /// </summary>
+        public Func<EndPoint, Socket>? CreateListenSocket { get; set; }
```
Can we preinitialize this with CreateDefaultListenSocket and make it non-nullable?

halter73 commented on src/Servers/Kestrel/Transport.Sockets/src/SocketTransportOptions.cs:
```diff
@@ -65,6 +68,47 @@ public class SocketTransportOptions
+            {
+                case FileHandleEndPoint fileHandle:
+                    return new Socket(
+                        new SafeSocketHandle((IntPtr)fileHandle.FileHandle, ownsHandle: true)
```
```suggestion
                        new SafeSocketHandle((IntPtr)fileHandle.FileHandle, ownsHandle: false)
```

Since we're not managing the lifetime anymore.

deanward81 replied:
Interestingly, when I do this the test case that uses a `FileHandleEndPoint` never returns - the shutdown of Kestrel hangs indefinitely (when calling `_transportManager.StopAsync()`) - presumably because the underlying socket isn't closed. In the case of the test this is because the underlying `Socket` I steal the handle from is static and I'm unable to manage its lifetime as it's consumed using `MemberData`.

I can absolutely fix the test case but I'm now dubious to change it because it's possible that Kestrel might hang if the caller doesn't clean-up the handle that they supposedly own. That seems like a bad place to be without much indication what the problem is.

deanward81 replied:
Thinking that adding a comment and leaving the `ownsHandle: true` might be a safer route to go here

halter73 replied:
This makes me think we should just go the `Func<EndPoint, Func<EndPoint, Socket>, Socket>` route since it's the most-flexible/least-tricky way to do this while continuing to manage `_socketHandle` lifetime. This is a pretty advanced API anyway. I don't like the idea of SafeHandle's finalizer controlling the lifetime of the socket.

halter73 replied:
Actually, it looks like Socket itself disposes the SafeSocketHandle if we claim ownership. It seems like we were doing a bunch of work for nothing. I think your change is good as is. Maybe just add a comment saying that `ownsHandle` indicates the [Socket itself will dispose the handle when it is Disposed](https://github.com/dotnet/runtime/blob/999c9c0bf1162e624533e2df58062d2811961fd3/src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.cs#L3270).

deanward81 committed:
Make `CreateListenSocket` non-nullable and initialize to `CreateDefaultListenSocket`

deanward81 committed:
Update test to use a time-based path for `UnixDomainSocketEndPoint`

halter73 commented:
New new proposed API based on https://github.com/dotnet/aspnetcore/pull/32827#discussion_r636519952

public class SocketTransportOptions
{
+    public Func<EndPoint, Func<EndPoint, Socket>, Socket> CreateListenSocket { get; set; }
}

halter73 reviewed (approved):
I now see that the SafeSocketHandle is disposed by Socket all along because we were specifying `ownsHandle: true`. This means the previous manual tracking of the SafeSocketHandle was redundant and there's no behavioral change. Or at least that's what it seems like. I'll step through the code to confirm before merging.

deanward81 committed:
Add clarifying comment

JamesNK commented:
@halter73 Static method is new as well for API review?

halter73 commented:
Yes. This is still `api-ready-for-review`. We shouldn't merge before it's approved.

BrennanConroy commented:
API review:
```
public static Socket CreateDefaultBoundListenSocket(EndPoint);
public Func<EndPoint, Socket> CreateBoundListenSocket { get; set; }
```

deanward81 commented:
@BrennanConroy currently the implementation of `CreateDefaultListenSocket` has not had `Bind` called on it - with the approved API it implies that is the case - presumably the implementation should be changed to do so?

deanward81 committed:
Tweak to match approved API - `CreateListenSocket` => `CreateBoundListenSocket`

Moves the call to `Socket.Bind` from the `SocketConnectionListener` to `SocketTransportOptions` and adds xmldoc detailing behaviour.

Also added additional comments and another test to validate the behaviour of `CreateDefaultBoundListenSocket` using different kinds of endpoints.

deanward81 commented:
I've updated the code to match the API review. Let me know if you want any additional tweaks @BrennanConroy

davidfowl commented:
@BrennanConroy @halter73 is this good?

halter73 commented:
Thanks @deanward81!

sebastienros commented:
Did we measure the perf impact of that change? Asking for a friend :/

halter73 commented:
@sebastienros We didn't measure the perf of this change, but I'd be shocked if this had a perf impact. This moves the creation of the *listen* socket (not the accepted socket) to a Func, but keeps the logic the same by default. Even if there was a perf hit from calling a Func, it's only called once per address at app startup.

halter73 commented:
#34769 would be more likely to affect `Connection: close` perf because it touches a per-connection code path. I don't think we explicitly perf tested that. However, it's basically just a refactoring, so it shouldn't impact perf.
